flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestValuesRuntimeFunctions.java

```diff
@@ -283,22 +293,29 @@ private abstract static class AbstractExactlyOnceSink extends RichSinkFunction<R
         private static final long serialVersionUID = 1L;
 
         protected final String tableName;
+        protected final DataType consumedDataType;
+        protected final DataStructureConverter converter;
+        protected transient ListState<Row> rawResultState;
+        protected transient List<Row> localRawResult;
 
-        protected transient ListState<String> rawResultState;
-        protected transient List<String> localRawResult;
-
-        protected AbstractExactlyOnceSink(String tableName) {
+        protected AbstractExactlyOnceSink(
+                String tableName, DataType consumedDataType, DataStructureConverter converter) {
             this.tableName = tableName;
+            this.consumedDataType = consumedDataType;
+            this.converter = converter;
         }
 
         @Override
         public void initializeState(FunctionInitializationContext context) throws Exception {
             this.rawResultState =
                     context.getOperatorStateStore()
-                            .getListState(new ListStateDescriptor<>("sink-results", Types.STRING));
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "sink-results",
+                                            ExternalSerializer.of(consumedDataType, true)));
```

Comment:
Is the `true` correct here? The input is not internal but `Row`, no?

Reply:
if `localRawResult` would be RowData this would be true and we won't need DataStructureConverter anymore

Reply:
right, initially I started with `RowData` and didn't remove the flag.

Reply:
I'm confused why did no tests fail with this?

Reply:
Maybe no tests, depend on this particular one.

Reply:
This would be tested only if
* RocksDB is used
* or state is restored

Reply:
Let's use the restore tests for deeply testing this utility then.